File: 101.symmetric-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isMirror(self, node1, node2):
        if not node1 and not node2:
            return True
        elif not node1 and node2:
            return False
        elif node1 and not node2:
            return False
        else:
            return node1.val == node2.val and \
                self.isMirror(node1.left, node2.right) and \
                self.isMirror(node1.right, node2.left)

    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if not root:
            return True
        return self.isMirror(root.left, root.right)

File: 101.symmetric-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* node1, TreeNode* node2) {
        if (!node1 && !node2) {
            return true;
        } else if (!node1 && node2) {
            return false;
        } else if (node1 && !node2) {
            return false;
        } else {
            return node1->val == node2->val &&
                isSymmetric(node1->left, node2->right) &&
                isSymmetric(node1->right, node2->left);
        }
    }

    bool isSymmetric(TreeNode* root) {
        if (!root) {
            return true;
        }
        return isSymmetric(root->left, root->right);
    }
};